### ISSUE TITLE
I can't subrepo working.

### ISSUE THREAD

**Issue by MTKnife** [ingydotnet/git-subrepo#449]
I apologize for bothering anyone, but I just can't seem to get the package to work, though it does look extremely useful.  The documentation could probably use improvement, especially some worked-out examples:  at the moment, I can't tell if I'm dealing with bugs or if I'm doing something fundamentally wrong.

We have a Git repo that contains a number of sub-projects.  We want to share part of it--the Python code--with another repo.  Therefore, I did a `git subrepo init`, creating a ".subrepo" file in the subdirectory in question, with the following content:

```
;
; This subdirectory is a git "subrepo", and this file is maintained by the
; git-subrepo command. See https://github.com/git-commands/git-subrepo#readme
;
[subrepo]
	remote = ssh://git@bitbucket.di2e.net:7999/oebdp/oe-bdp.git
	branch = python-subrepo
	commit = 
	method = rebase
	cmdver = 0.4.1
```

At one point, I edited it to change the branch name, but I've since changed it back.

The subdirectory I need to push we'll call, for the present purposes, "python-subdir", and the most up-to-date version of the Python code we need is in a branch that we'll call "my-branch".  Some of its code may not have been merged into "master" yet (the repo host appears to be down at the moment, so I can't check easily); however, "master" does have the ".subrepo" file, though no branches besides that and "my-branch" have it.  However, "my-branch" has been rebased on the current master, and pushed that way, so there are no merge conflicts.

The docs can be a little confusing when it comes to figuring out whether a given parameter refers to the source subdirectory vs. the target subdirectory, and the subrepo branch vs. the branch I'm trying to push  to the subrepo.  However, I _think_ this is the command I should be typing next:

`git subrepo push python-subdir -b my-branch`

What I get in response to that command is:

`git-subrepo: There are new changes upstream, you need to pull first.`

There absolutely are not changes upstream--as I said, I rebased the source branch on the master, then pushed it--so this baffles me.  I've tried doing a `git subrepo clean`, as well as removing artifacts manually, but I can't get anything to work.

Any insights would be appreciated.

**Comment by MTKnife** [comment]
Anyone?  This can't be the first time something similar has happened.

**Comment by admorgan** [comment]
I think there is some confusion. I agree the docs can be a bit confusing and I am happy to take patches to assist with that.

git subrepo init sets up a new repository. As states "Note: You will need to create the empty upstream repo and push to it on your own, using git subrepo push <subdir>." So for example if you were wanting to split your python code out to share on github you would have to create an empty github remote, then run git subrepo init <your subdir> -r <github location>, then you can run git subrepo push.

It would be strange to use the branch option as that refers to the remote branch and not the local one. It is double weird to use the branch option and then use a different branch in the push command.

**Comment by MTKnife** [comment]
So if I have two repos I want to keep in sync, I would end up having sub-directories in both repos, and a third repo for the subrepo push?

**Comment by admorgan** [comment]
I think we are still having a communication problem. Lets establish some vocabulary. A project is  a git repository, a subrepo is a directory that contains the content of another repository.

If you have a project that uses subrepos it may have the pattern of:
project/
           subrepo1/
           subrepo2/
           Makefile
           tool.c
           python-tools/

In this case if you wanted to make python-tools a new repository somewhere else you would need to create an empty repository at that location then call git init <url for your empty repository> python-tools && git push python-python tools.

If you want to get the most recent changes in subrepo1 you would do a git subrepo pull subrepo1
If you made changes in a commit to subrepo1 and you want to share those with the original repository then you would do a git push subrepo1.

I hope this is helpful. I am happy to answer any additional questions you may have.

**Comment by MTKnife** [comment]
Yes, that much I got.  It was just that the push wasn't working.  It seems to have worked now, though. Thanks.

Oddly, this time when I used the "init", method was set to "merge" rather than "rebase".  Any idea why?

**Comment by admorgan** [comment]
Merge is the default. I am not sure why it would have been rebase unless you passed --rebase as an option.
